### PR TITLE
Add logic to determine faster route

### PR DESCRIPTION
**Pull request by abhishek1508**
## Description

Add logic to use user current location to request for faster route. Compare new route with current route to determine if it is faster [#221](https://github.com/mapbox/navigation-sdks/issues/221)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Add logic to use user current location to request for faster route. Compare new route with current route to determine if it is faster.

### Implementation

Add logic to use user current location to request for faster route. Compare new route with current route to determine if it is faster.

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSession.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.mapbox.navigation.base.session
+
+interface NavigationSession {
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSession.kt]: Do we need to expose this in the `base` module? If we only want to use it for internal purposes in `core, let's keep it this way.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.mapbox.navigation.base.session
+
+interface NavigationSessionObserver {
```
Comment (LukasPaczos) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: Same as above.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 3]
```diff
@@ -106,9 +112,16 @@ class MapboxNavigation(
...
-    private val navigationSession = NavigationSession(context)
+    private val navigationSession = MapboxNavigationSession(context)
     private val internalRoutesObserver = createInternalRoutesObserver()
     private val internalOffRouteObserver = createInternalOffRouteObserver()
+    private val internalLocationObserver = createInternalLocationObserver()
+    private val internalRouteProgressObserver = createInternalRouteProgressObserver()
+    private val internalNavigationSessionObserver = createInternalNavigationSessionObserver()
+    private val fasterRouteTimer: MapboxTimer
+
+    private lateinit var enhancedLocation: Location
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: We don't need this field and the observer that assigns it, we can always call `TripSession#tripSession.getEnhancedLocation()`.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 4]
```diff
@@ -106,9 +112,16 @@ class MapboxNavigation(
...
+    private lateinit var routeProgress: RouteProgress
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: Same as above.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 5]
```diff
@@ -401,6 +458,26 @@ class MapboxNavigation(
         }
     }
 
+    private fun createInternalNavigationSessionObserver() = object : NavigationSessionObserver {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: Do we need this observer? We can keep providing faster route updates even if we are not navigating. This can still be useful to keep updating the UI and the decision to do so can be made by the user.

Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: Yes. To me it doesn't make sense to make faster route request unless you start active guidance. We already provide alternative route at that moment to allow the user to make the decision about which route is faster. I think what you would probably want here is to fire route refresh the moment you make a route request. `RouteRefresh` gives you better traffic updates, ETA's etc. which you might wanna reflect on all the routes. If you make faster route request with the same origin there is almost no chance that we get any different route. 

Hence, based on these discussions we agreed to query for faster route only when you have started navigation. This may change later, as I heard that there are talks about emitting `RouteProgress` during free drive.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSession.kt (hunk 1 above)
Comment (abhishek1508) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSession.kt]: I thought it would be a good idea to expose this because there are situations where app developers want to listen to state changes as to whether we are actively navigating or not. Exposing for such observer to state changes help them to directly subscribe to it rather than hacking their way out to find out when are we navigating or not

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt (hunk 2 above)
Comment (abhishek1508) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: Same as above

**Comment by abhishek1508** [comment]
> Good progress 👍
> 
> How about `DirectionsSession` exposes a callback like in #2455, but instead of calling it `FaterRouteCallback`, we call it `AlternativeRouteObserver`.
> The session still maintains the timer and makes periodic requests that invoke the observers.
> 
> `MapboxNavigation` can subscribe to this event and based on the route progress determine whether the alternative route is actually faster. If it is, it notifies `FaterRouteObservers` that now live and are maintained in `MapboxNavigation`.
> 
> This way, we are decreasing the coupling between the 2 classes and simplifying the whole flow.

I had a great discussion with @Guardiola31337 last week on 2 different approaches to implement faster route detection. We explored the possibility of having the `Faster Route Timer` in `DirectionsSession`. The problem is to make a faster route request you use the same API as route request. The API needs `RouteOptions` which needs updated user current location. DirectionsSession doesn't have access to users current location. We don't want to expose methods to DirectionSession that maintains users current location or make DirectionsSession observe location updates. The architecture was implemented to make sure that `MapboxNavigation` will be the intermediary b/w `Directions` and `Trips` and any feature that needs access to properties from these 2 different modules will be handled in `MapboxNavigation`. Hence, the timer cannot live in `DirectionsSession`.

The implementation above is the first iteration towards `FasterRoute`. I plan to take out the implementation details from `MapboxNavigation` so that it is light in the next iteration. Also, I would rather not use the term `AlternativeRouteObserver` as I feel that may be confusing from client's perspective. I say this because there is already this feature where we allow users to specify if they want alternative routes. and draw them accordingly. We then have offRoute observers and callbacks that inform users about `onRoutesChanged`. Having specific terms for features makes it easy to understand what is one doing. Wdyt?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: I see the usefulness. I'm just worried that by exposing those states, we're exposing internal mechanisms and tying ourselves to certain expectations that could not be changed later on with additional features. After all, we might be in the internal "Free Drive" mode, but if there are no tiles on the device, enhanced location will be equal to raw location and user's shouldn't have to worry about it etc. Same for `Active Guidance`, user can change to the Navigation UI view whenever convenient, we don't have to be in the internal `Active Guidance` state. The states bring value to us (like billing) but might not to the end-user. WDYT @mapbox/navigation-android?

Regardless, these observers can live in `core` only and be exposed from there, no modules depend on them.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSession.kt (hunk 1 above)
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSession.kt]: See https://github.com/mapbox/mapbox-navigation-android/pull/2462#discussion_r378821137 for my thoughts. But whether or not we expose the observers, this interface looks unnecessary to me. Simply exposing register/unregister methods from `MapboxNavigation` should be enough.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 5 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: `DirectionsSession` should preferably be separated as much as possible from the `TripSession`. By depending on `NavigationSession` to request a faster route, we are again tying them together. We can let the user make a choice of whether or not to accept a faster route if `TripSession` is stopped.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 6]
```diff
@@ -61,6 +53,39 @@ class MapboxDirectionsSession(
...
+        if (routes.isEmpty()) {
+            return
+        }
+        routeOptions?.let { options ->
+            val origin = Point.fromLngLat(currentLocation.longitude, currentLocation.latitude)
+            val coordinates = options.coordinates().apply {
+                removeAt(0)
+                add(0, origin)
+            }
+            options.toBuilder().coordinates(coordinates)
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]:
```suggestion
            val adjustedOptions = options.toBuilder().coordinates(coordinates).build()
```

This might not be enough though. We need to remove all waypoints that have already been visited, adjust their names, radiuses etc. Basically the same logic that needs to happen in
https://github.com/mapbox/mapbox-navigation-android/blob/83da15a0052645546da07b26beef9e8eaa1ec229/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L404-L453

just with a different request method. In this case, maybe it'd be better to compile new options in the `MapboxNavigation`?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteRequestCallback.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
+package com.mapbox.navigation.core.fasterroute
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.navigation.base.route.Router
+
+interface FasterRouteRequestCallback {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteRequestCallback.kt]: Could we just reuse `RoutesRequestCallback`?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt [hunk 8]
```diff
@@ -107,72 +99,6 @@ class MapboxDirectionsSessionTest {
         verify { router.cancel() }
     }
 
-    @Test
-    fun fasterRoute_timerStartedOnce() {
```
Comment (LukasPaczos) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt]: Can we recreate these test in `MapboxNavigation`?

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt (hunk 2 above)
Comment (abhishek1508) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: What if the `MapboxService` or `TripNotification` wants to listen to the state changes? I think the existing telemetry code from legacy SDK also performs some actions based on active navigation. How would `metrics` module which contains `telemetry` in the new v1.0 listen to these navigation state changes?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteRequestCallback.kt (hunk 7 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteRequestCallback.kt]: I am afraid not. 👀 at 
```
fun onRouteReady(route: DirectionsRoute): Boolean
```

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt (hunk 8 above)
Comment (abhishek1508) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt]: I am working on it

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 6 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: This is a good point. Will fix it

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: I think the metrics module is being moved to core in https://github.com/mapbox/mapbox-navigation-android/pull/2445. But regardless, if we need to push a state from `MapboxNavigaiton` to an upstream module, it means that the module is probably setup incorrectly because it has dependencies on the downstream logic.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 9]
```diff
@@ -61,6 +53,39 @@ class MapboxDirectionsSession(
...
+            router.getRoute(options, object : Router.Callback {
+                override fun onResponse(routes: List<DirectionsRoute>) {
+                    val isRouteFaster = fasterRoutesRequestCallback.onRouteReady(routes[0])
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: I think `DirectionsSession` should just deliver the route to the client (`MapboxNavigation`) and that client should determine if the route is faster and notify the subscribers. We don't need to cycle back to the `DirectionsSession`.
This basically means that `FasterRouteCallback` could be maintained and invoked from `MapboxNavigation` instead.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 5 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: I don't think we are tying `DirectionsSession` and `TripSession` here. We are just invoking some functionality from `directionSession` only if certain changes happen in the `navigationSession`. If not now, there will be n number of use cases that will come up later where features would want to know the current state of navigation. I have experienced that as a SDK user many times. I don't think exposing navigation states is same as exposing internal mechanisms. The user/end client already knows what state they are in - `FREE_DRIVE`, `ACTIVE_GUIDANCE`, `IDLE`. All navigation state is doing is just giving them an opportunity to listen to those changes.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt (hunk 2 above)
Comment (abhishek1508) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: Then we need to think about it. Navigation State changes is something that should be available to everyone. The architecture might have to be modified a little bit to address this [issue](https://github.com/mapbox/navigation-sdks/issues/231). We may end up doing some stuff in background which always runs even if the activity is force closed from recents. I know the legacy code in this case needed to know active guidance state in order to decide when to close this background job.
We need to consider how we want to expose the state changes both to the different modules in the SDK and end clients as well

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 9 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: If you look at this https://github.com/mapbox/mapbox-navigation-android/pull/2455
> ```
> private fun requestFasterRoute(routeOptions: RouteOptions) {
>        if (routes.isEmpty()) {
>            return
>        }
>        router.getRoute(routeOptions, object : Router.Callback {
>            override fun onResponse(routes: List<DirectionsRoute>) {
>                val route = routes[0]
>                if (isRouteFaster(route)) {
>                    fasterRouteObservers.forEach { it.onFasterRouteAvailable(route) }
>                }
>            }
>
>            override fun onFailure(throwable: Throwable) {
>                // do nothing
>            }
>
>            override fun onCanceled() {
>                // do nothing
>            }
>        })
>    }
> ```

```
if (isRouteFaster(route)) {
       fasterRouteObservers.forEach { it.onFasterRouteAvailable(route) }
}
```

I have a `fasterRouteObserver` to which the end user subscribes to. This needs access to faster route. Hence, this is done this way

**Comment by abhishek1508** [comment]
I don't see any other outstanding comments [here](https://github.com/mapbox/mapbox-navigation-android/pull/2455).
@Guardiola31337 @LukasPaczos would be great if you could approve. I would like to merge this so that this code is then accessible in this ticket and it makes it easy for everyone to understand how is the entire faster route tied together

**Comment by LukasPaczos** [comment]
Capturing from multiple threads above:
- I'm okay with exposing the states if you feel like it's the right thing to do now. I can see the value, we'll just need to be extra careful.
- I still think that the faster route timer should run even if we are not in active guidance and leave the decision of whether or not to update the route to the user.

**Comment by abhishek1508** [comment]
> * I'm okay with exposing the states if you feel like it's the right thing to do now. I can see the value, we'll just need to be extra careful.

@LukasPaczos I would like to give some time to think on what are the possible other ways to expose navigationState in general. I think this is good for first iteration. We should definitely have a broader discussion around this in a separate follow up ticket that is more targeted to `NavigationState` rather than faster route. Will create a ticket for it.

**Comment by LukasPaczos** [comment]
Sounds good, thanks @abhishek1508.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 9 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: Refs https://github.com/mapbox/mapbox-navigation-android/pull/2455/files#r379006808.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 5 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt]: Created [ticket](https://github.com/mapbox/navigation-sdks/issues/242) to explore how we may want to modify it in the next iteration.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt (hunk 2 above)
Comment (abhishek1508) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/session/NavigationSessionObserver.kt]: Created ticket to explore how we may want to modify it in the next iteration.

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt [hunk 10]
```diff
@@ -7,13 +7,16 @@ import com.mapbox.navigation.base.typedef.RoundingIncrement
...
+const val DEFAULT_FASTER_ROUTE_DETECTOR_INTERVAL = 2 * 60 * 1000L // 2 minutes
 
 class NavigationOptions private constructor(
     @RoundingIncrement private val roundingIncrement: Int,
     @TimeFormatType private val timeFormatType: Int,
     private val navigatorPollingDelay: Long,
     private val distanceFormatter: DistanceFormatter?,
-    private val onboardRouterConfig: MapboxOnboardRouterConfig?
+    private val onboardRouterConfig: MapboxOnboardRouterConfig?,
+    private val enableFasterRouteDetection: Boolean,
```
Comment (kmadsen) [inline libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt]: Can we remove this from a big NavigationOptions constructor?

Ideally we should be able to enable or disable faster route detection independently. Customers will want to be able to give their users the ability to enable/disable the feature without requiring the trip to start over.

**Comment by Guardiola31337** [comment]
Capturing from internal discussion with @abhishek1508 that we're going to try the following:

- Move `NavigationSession` back to `libnavigation-core` (for now)
- Remove unnecessary back and forth cycle between `MapboxNavigation` and `DirectionsSession`
- Reuse `RoutesRequestCallback`
- Run faster route timer independently of the `NavigationSession.State`
- Link faster route timer lifecycle to observer `register` / `unregister`

cc @LukasPaczos

File: libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt (hunk 10 above)
Comment (Guardiola31337) [reply libnavigation-base/src/main/java/com/mapbox/navigation/base/options/NavigationOptions.kt]: That's a good point. For now after discussing with @abhishek1508 we're going implement 👇 https://github.com/mapbox/mapbox-navigation-android/pull/2462#issuecomment-586020114

> Link faster route timer lifecycle to observer `register` / `unregister`

so it can be removed from here and enabled / disabled as part of `register` and `unregister` APIs. For `fasterRouteDetectorInterval` we can leave it as is (initialized at `MapboxNavigation` construction time and attached to its lifecycle). If we feel like clients want to change the interval dynamically we can think about it later and iterate in a future PR. How does that sound @kmadsen?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.mapbox.navigation.core.fasterroute
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.trip.model.RouteProgress
+
+object FasterRouteDetector {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt]: This can be internal.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt (hunk 8 above)
Comment (LukasPaczos) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt]: Still valid.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 12]
```diff
@@ -63,6 +50,29 @@ class MapboxDirectionsSession(
         })
     }
 
+    override fun requestFasterRoute(
+        adjustedRouteOptions: RouteOptions,
+        routesRequestCallback: RoutesRequestCallback
+    ) {
+        if (routes.isEmpty()) {
+            return
+        }
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: How about we call this method `requestAlternativeRoutes`? This would better describe what the method is actually doing and might be reused for other purposes in the future.

If we'd do that, we can move `if (routes.isEmpty())` check before invoking this method instead of having it here.

Just an idea, I'm fine with the current setup as well.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 13]
```diff
@@ -63,6 +50,29 @@ class MapboxDirectionsSession(
...
+        this.routeOptions = adjustedRouteOptions
+        router.getRoute(adjustedRouteOptions, object : Router.Callback {
+            override fun onResponse(routes: List<DirectionsRoute>) {
+                routesRequestCallback.onRoutesReady(routes)
+            }
+
+            override fun onFailure(throwable: Throwable) {
+                // do nothing
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: let's invoke the callback here

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 14]
```diff
@@ -63,6 +50,29 @@ class MapboxDirectionsSession(
...
+            }
+
+            override fun onCanceled() {
+                // do nothing
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: let's invoke the callback here

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/DirectionsSession.kt [hunk 15]
```diff
@@ -12,16 +11,14 @@ internal interface DirectionsSession {
 
     fun requestRoutes(routeOptions: RouteOptions, routesRequestCallback: RoutesRequestCallback)
 
+    fun requestFasterRoute(adjustedRouteOptions: RouteOptions, routesRequestCallback: RoutesRequestCallback)
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/DirectionsSession.kt]: Can we add a doc here that this method does the same thing that `requestRoutes`, but ignores the result and it's up to the consumer to take the action with the route?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+
+    fun isRouteFaster(newRoute: DirectionsRoute, routeProgress: RouteProgress): Boolean {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt]: Is the logic for faster route going to be included in a separate PR?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 12 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: The idea to reuse it for other purposes in the future sounds good to me. 

> Also, I would rather not use the term AlternativeRouteObserver as I feel that may be confusing from client's perspective. I say this because there is already this feature where we allow users to specify if they want alternative routes. and draw them accordingly. We then have offRoute observers and callbacks that inform users about onRoutesChanged. Having specific terms for features makes it easy to understand what is one doing. Wdyt?

I agree with @Guardiola31337 here. I wouldn't want to use the term `Alternative` here.
I would leave it as it is for now. Whenever we feel this can be reused we can revisit and think about renaming.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt (hunk 16 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt]: Yeah this requirement was not captured in this ticket.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 17]
```diff
@@ -63,6 +51,33 @@ class MapboxDirectionsSession(
         })
     }
 
+    override fun requestFasterRoute(
+        adjustedRouteOptions: RouteOptions,
+        routesRequestCallback: RoutesRequestCallback
+    ) {
+        if (routes.isEmpty()) {
+            return
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: We should call `routesRequestCallback#onRoutesRequestFailure` here.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt [hunk 18]
```diff
@@ -109,82 +99,6 @@ class MapboxDirectionsSessionTest {
...
-        callback.onResponse(routes)
-
-        clearMocks(router)
-        delayLambda()
-        session.cancel()
-        verify(exactly = 1) { router.cancel() }
-    }
-
-    @Test
-    fun fasterRoute_canceledByNewRequest() {
```
Comment (LukasPaczos) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSessionTest.kt]: We need to recreate this test - when a faster route request is sent, it should be canceled whenever a new, main route is requested/set.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 17 above)
Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: I think it should rather be `routesRequestCallback#onRoutesRequestCanceled` because failure means the request was made and failed for some reason; whereas in this case we don't even make the request. Wdyt?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt (hunk 16 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/fasterroute/FasterRouteDetector.kt]: Let's cut a follow up ticket then @abhishek1508 👍

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 17 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: Sounds good 👌 @abhishek1508

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt [hunk 19]
```diff
@@ -42,12 +69,19 @@ class MapboxNavigationTest {
 
     @Before
     fun setUp() {
+
```
Comment (Guardiola31337) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt]: NIT unnecessary new line

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt [hunk 20]
```diff
@@ -127,4 +172,97 @@ class MapboxNavigationTest {
 
         verify(exactly = 1) { tripSession.unregisterAllVoiceInstructionsObservers() }
     }
+
+    @Test
+    fun fasterRoute_timerStarted() {
```
Comment (Guardiola31337) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt]: Minor thing

Normally is good to structure the tests following the triple A 👀 https://speakerdeck.com/guardiola31337/elegant-unit-testing-mobilization-2016?slide=40 http://wiki.c2.com/?ArrangeActAssert
If the `A`rrange part gets too long we should create factory or builder methods.

This applies to the rest of the tests 👇

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt [hunk 21]
```diff
@@ -63,6 +51,33 @@ class MapboxDirectionsSession(
...
+        }
+        this.routeOptions = adjustedRouteOptions
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: It seems we're changing the "global" route options when `requestFasterRoute` is called. Is that intentional? Could that be a problem?

Comment (abhishek1508) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: Thanks @Guardiola31337 
This was a very good catch. 

The question I would like to ask is do we want to change the global `routeOptions` if the user selects a faster route? 

It is different from `requestRoute` where user doesn't have the option to accept the route. This global `routeOptions` should only be changed if the user agrees to accept the faster route.

**Comment by abhishek1508** [comment]
@Guardiola31337 `fasterRoute_timerStartedOnce` should not be concern of `MapboxNavigation` or `MapboxDirectionSession`. All they should do is call `start` on the timer and timer should be responsible to check if it is being started once or multiple times. `MapboxTimerTest` doesn't exist at the moment, hence I am going to cut a ticket for it.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt (hunk 21 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/MapboxDirectionsSession.kt]: Synced up internally and we're going to get the route options from the `DirectionsRoute`, set them when routes are set and remove the explicit set from `requestRoutes` and `requestFasterRoute`.

**Comment by Guardiola31337** [comment]
👀 @cafesilencio @evabishchevich @JunDai this PR brings some insights and thoughts on what we were discussing today. Please read the threads comments when you have a chance.

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt (hunk 20 above)
Comment (Guardiola31337) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt]: NIT This is still valid
